# DS 3000 - Assignment 2

**Student Name**: Bella Chin

**Date**: 09/12/2024


### Submission Instructions
Submit this `ipynb` file to canvas.

The `ipynb` format stores outputs from the last time you ran the notebook.  (When you open a notebook it has the figures and outputs of the last time you ran it too).  To ensure that your submitted `ipynb` file represents your latest code, make sure to give a fresh run `Kernel > Restart & Run All` just before uploading the `ipynb` file to Canvas.

### Academic Integrity

**Writing your homework is an individual effort.**  You may discuss general python problems with other students but under no circumstances should you observe another student's code which was written for this assignment, from this year or past years.  Pop into office hours or DM us in MS Teams if you have a specific question about your work or if you would like another pair of eyes or talk through your code.

Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment. You do not need to cite the official python documentation.

**Documentation / style counts for credit**  Please refer to the Pep-8 style, to improve the readability and consistency of your Python code. For more information, read the following article [How to Write Beautiful Python Code With PEP 8](https://realpython.com/python-pep8/) or ask your TA's for tips.

**NOTE:<span style='color:red'> Write python expressions to answer ALL questions below and ensure that you use the `print()` function to display the output.</span>** Each question should be answered in a new code cell. For example, your solution for question 1.1 should be in a different code cell from your solution for question 1.2.

# Question 1: 

Load the attached data into a numpy array and display the dimensions of the data:

In [3]:
#Write your python code to answer question 1. You can add new cells below as needed.

In [5]:
import numpy as np

# Load the data from CSV file
file_name = '/Users/snazzy/Desktop/DS3000/birthwt.csv'

#Load data from CSV file with header
data_with_header = np.genfromtxt(file_name, delimiter = ',', dtype = None, encoding = None)

#Load data without header
data = np.loadtxt(file_name, 
                  delimiter=',',
                  skiprows=1, dtype = int) 


#Create header dictionary
headers = data_with_header[0]
header_dict = {header: index for index, header in enumerate(headers)} 
#used w3resource for how to convert numpy array to dictionary with indicies as keys
dimensions = data.shape

# Display the dimensions of the data.
print(dimensions)
print("The dimensions of the data are", dimensions[0], "rows and", dimensions[1], "columns.")

(189, 10)
The dimensions of the data are 189 rows and 10 columns.


## Question 2: 
What is the age range of mothers in this dataset.
Note: the range should include: the minimum, average and maximum age. Ensure that the ages are represented as whole numbers.

In [11]:
# Get the index of the 'age' column using the header dictionary
age_column_index = header_dict["age"]

# Extract the 'age' column from the data
age_column = data[:, age_column_index]

# Display the extracted age column
print("Age column:", age_column)


Age column: [19 33 20 21 18 21 22 17 29 26 19 19 22 30 18 18 15 25 20 28 32 31 36 28
 25 28 17 29 26 17 17 24 35 25 25 29 19 27 31 33 21 19 23 21 18 18 32 19
 24 22 22 23 22 30 19 16 21 30 20 17 17 23 24 28 26 20 24 28 20 22 22 31
 23 16 16 18 25 32 20 23 22 32 30 20 23 17 19 23 36 22 24 21 19 25 16 29
 29 19 19 30 24 19 24 23 20 25 30 22 18 16 32 18 29 33 20 28 14 28 25 16
 20 26 21 22 25 31 35 19 24 45 28 29 34 25 25 27 23 24 24 21 32 19 25 16
 25 20 21 24 21 20 25 19 19 26 24 17 20 22 27 20 17 25 20 18 18 20 21 26
 31 15 23 20 24 15 23 30 22 17 23 17 26 20 26 14 28 14 23 17 21]


In [7]:
# Get the index the age is at
age_index = header_dict['age']

# Obtain the mother's ages
mother_ages = data[:, age_index]

#Find the min, max, and average of the ages
min_mother_age = data[:, age_index].min()
max_mother_age = data[:, age_index].max()
avg_mother_age = int(data[:, age_index].mean())

# Display the minimum, maximum, and average mother age.
print("The minimum age of mothers in this dataset is", min_mother_age, 
      ".The maximum age of mothers in this dataset is", max_mother_age, 
      ".The average age of mothers in this dataset is", avg_mother_age ,".")


The minimum age of mothers in this dataset is 14 .The maximum age of mothers in this dataset is 45 .The average age of mothers in this dataset is 23 .


# Question 3:
How many mothers smoked during pregnancy?

In [9]:
# Get the index of smoke status from the header dictionary
smoke_index = header_dict['smoke']

# Get the status of mother's who smoked or did not smoke
smoke_status = data[:, smoke_index]

# Go through the smoke statuses and filter if mothers smoked
count = 0
for smoke in smoke_status:
    if (smoke == 1):
        count += 1
        
# Display output of mothers who smoked
print(count, "mothers have smoked during pregnancy.")

74 mothers have smoked during pregnancy.


# Question 4:
How many mothers had a history of hypertension?

In [11]:
# Get the index of hypertension from the header dictionary
hyp_index = header_dict['ht']

# Get whether mothers have history of hypertension or do not have history of hypertension.
hyp_hist = data[:, hyp_index]

# Filter the mothers who have a history of hypertension.
count = 0
for hyp in hyp_hist:
    if (hyp == 1):
        count += 1

# Display the motheres who have a history of hypertension.
print(count, "mothers had a history of hypertension.")

12 mothers had a history of hypertension.


# Question 5:
How many mothers have 1 or more premature births?

In [13]:
# Get the index of premature births from the header dictionary
ptl_index = header_dict['ptl']

# Get the number of premature births each mother had.
ptl_num = data[:, ptl_index]

# Filter the mothers who had one or more premature births
count = 0
for ptl in ptl_num:
    if (ptl >= 1):
        count += 1

# Display the mothers who had 1 or more premature births
print(count, "mothers had 1 or more premature births.")

30 mothers had 1 or more premature births.


# Question 6:
Create a function called convert_birthweight() that: 1) takes grams as input, 2) converts grams to pounds and ounces, and 3) return the converted value. Use the convert_birthweight() function to convert the birth weight to pounds and ounces in the questions 7-11 below.

In [15]:
# User defined function that helps convert the birth weight in grams into pounds and ounces
def convert_birthweight(grams):
    pounds = grams * 0.00220462
    wholenum_pounds = int(pounds)
    ounces = (pounds - wholenum_pounds) * 16
    return wholenum_pounds, ounces
    

# Question 7:
What is the average birth weight of babies whose mothers have a history of premature birth? 

In [17]:
# Get the index of birth weight from the header dictionary
bwt_index = header_dict['bwt']

# Get the birtheight data using the index
bwt_nums = data[:, bwt_index]

# Find the indicies where mothers had premature birth
indices = np.where(ptl_num >= 1)[0] #geeksforgeeks numpy where

# Filter the data and get the birth weight of those who had premature births
ptl_bwt = bwt_nums[indices]

# Find the mean of the premature birth weights
avg_ptl_bwt = ptl_bwt.mean()

# Convert the birth weights into pounds and ounces.
wholenum_pounds, ounces = convert_birthweight(avg_ptl_bwt)

# Display the average birth weight in pounds and ounces.
print("The average birth weight of babies whose mothers have a history of premature birth is" ,
      wholenum_pounds, "pounds and", round(ounces, 2), "ounces.")

The average birth weight of babies whose mothers have a history of premature birth is 5 pounds and 10.99 ounces.


# Question 8:
What is the minimum, average and maximum birth weight of babies whose mother did not smoke during pregnancy and do not have a history of hypertension?

In [19]:
# Find the indices where mothers did not smoke and do not have a history of hypertension
indices = np.where((smoke_status != 1) & (hyp_hist != 1))[0] 
#sparkbyexamples numpy where() multiple conditions

# Get the data of those who did not smoke and do not have history of hypertension using indicies
no_smk_ht = bwt_nums[indices]

# Find min, max, and avg or the data
min_no_smk_ht = no_smk_ht.min()
max_no_smk_ht = no_smk_ht.max()
avg_no_smk_ht = no_smk_ht.mean()

# Convert the numbers into pounds and ounces
min_pounds, min_ounces = convert_birthweight(min_no_smk_ht)
max_pounds, max_ounces = convert_birthweight(max_no_smk_ht)
avg_pounds, avg_ounces = convert_birthweight(avg_no_smk_ht)

# Display the min, max, and avg of the birth weight of mothers who did not smoke and did not have history of hypertension
print("The minimum birth weight of babies whose mother did not smoke during pregnancy and do not",
"have a history of hypertension is", min_pounds, "pounds and", round(min_ounces,2),
      "ounces. The maximum birth weight is",
     max_pounds, "pounds and", round(max_ounces,2), "ounces.The average birth weight is"
      , avg_pounds, "pounds and", round(avg_ounces,2), "ounces.")

The minimum birth weight of babies whose mother did not smoke during pregnancy and do not have a history of hypertension is 2 pounds and 4.01 ounces. The maximum birth weight is 11 pounds and 0.02 ounces.The average birth weight is 6 pounds and 13.01 ounces.


# Question 9:
What is the minimum, average and maximum birth weight of babies whose mothers smoked during pregnancy?

In [21]:
# Get the indicies where mothers smoked
indices = np.where((smoke_status == 1))[0]

# Get the birth weight data of mothers who smoked
smk_bwt = bwt_nums[indices]

# Find the min, max, and avg of birth weight of mothers who smoked
min_smk_bwt = smk_bwt.min()
max_smk_bwt = smk_bwt.max()
avg_smk_bwt = smk_bwt.mean()

# Convert each to pounds and ounces by calling the function
min_smk_pounds, min_smk_ounces = convert_birthweight(min_smk_bwt)
max_smk_pounds, max_smk_ounces = convert_birthweight(max_smk_bwt)
avg_smk_pounds, avg_smk_ounces = convert_birthweight(avg_smk_bwt)

# Display the min, max, and avg birth weight of mothers who smoked
print("The minimum birth weight of babies whose mothers smoked during pregnancy is",
      min_smk_pounds, "pounds and", round(min_smk_ounces, 2), "ounces. The maximum is",
      "birth weight of babies whose mothers smoked during pregnancy is",
      max_smk_pounds, "pounds and", round(max_smk_ounces, 2), "ounces. The average is",
      avg_smk_pounds, "pounds and", round(avg_smk_ounces, 2), "ounces.")

The minimum birth weight of babies whose mothers smoked during pregnancy is 1 pounds and 9.01 ounces. The maximum is birth weight of babies whose mothers smoked during pregnancy is 9 pounds and 5.49 ounces. The average is 6 pounds and 1.78 ounces.


# Question 10: 
What is the minimum, average and maximum birth weight of babies whose mothers had a history of hypertension? 

In [23]:
# Find the indices where mothers had a history of hypertension
indices = np.where((hyp_hist == 1))[0]

# Get the birth weight data of mothers who had hypertension
hyp_bwt = bwt_nums[indices]

# Find the min, max, and avg birth weight of mothers who had a history of hypertension
min_hyp_bwt = hyp_bwt.min()
max_hyp_bwt = hyp_bwt.max()
avg_hyp_bwt = hyp_bwt.mean()

# Convert the min, max, and avg birth weight of mothers who had a history of hypertension
min_hyp_pounds, min_hyp_ounces = convert_birthweight(min_hyp_bwt)
max_hyp_pounds, max_hyp_ounces = convert_birthweight(max_hyp_bwt)
avg_hyp_pounds, avg_hyp_ounces = convert_birthweight(avg_hyp_bwt)

# Display the min, max, and avg birth weight of motehrs who had a history of hypertension
print("The minimum birth weight of babies whose mothers had a history of hyperension is",
      min_hyp_pounds, "pounds and", round(min_hyp_ounces, 2), "ounces. The maximum",
      "birth weight of babies whose mothers had a history of hyperension is",
      max_hyp_pounds, "pounds and", round(max_hyp_ounces, 2), "ounces. The average is",
      avg_hyp_pounds, "pounds and", round(avg_hyp_ounces, 2), "ounces.")

The minimum birth weight of babies whose mothers had a history of hyperension is 2 pounds and 8.04 ounces. The maximum birth weight of babies whose mothers had a history of hyperension is 8 pounds and 5.69 ounces. The average is 5 pounds and 9.48 ounces.


# Question 11:
What is the minimum, average and maximum birth weight of babies whose mothers smoked during pregnancy and had a history of hypertension? 

In [25]:
# Find indicies where mothers smoked and had a history of hypertension
indices = np.where((smoke_status == 1) & (hyp_hist == 1))[0] 

# Get birth weight data of mothers who smoked and had a history of hypertension
smk_ht = bwt_nums[indices]

# Get the min, max, and avg birth weight of mothers who smoked and had a history of hypertension
min_smk_ht = smk_ht.min()
max_smk_ht = smk_ht.max()
avg_smk_ht = smk_ht.mean()

# Convert the min, max, and avg birth weight into pounds and ounces by calling the function
min_smk_ht_pounds, min_smk_ht_ounces = convert_birthweight(min_smk_ht)
max_smk_ht_pounds, max_smk_ht_ounces = convert_birthweight(max_smk_ht)
avg_smk_ht_pounds, avg_smk_ht_ounces = convert_birthweight(avg_smk_ht)

# Display the min, max, and avg birth weight of mothers who smoked and had history of hypertension during pregnancy
print("The minimum birth weight of babies whose mother smoked during pregnancy and had",
"a history of hypertension is", min_smk_ht_pounds, "pounds and", round(min_smk_ht_ounces,2),
      "ounces. The maximum birth weight is",
     max_smk_ht_pounds, "pounds and", round(max_smk_ht_ounces,2), 
      "ounces.The average birth weight is"
      , avg_smk_ht_pounds, "pounds and", round(avg_smk_ht_ounces,2), "ounces.")

The minimum birth weight of babies whose mother smoked during pregnancy and had a history of hypertension is 2 pounds and 8.04 ounces. The maximum birth weight is 8 pounds and 4.49 ounces.The average birth weight is 5 pounds and 10.34 ounces.


# Question 12:
Compare and contrast the results from 7-11 above and summarize your findings.
Do you see any patterns in the birth weight of babies whose mother did not have any history of hypertension and/or did not smoke during pregnancy versus those who did? Ensure that you reference specific birth weights to support your explanation.

The average birth weight of babies whose mother did not smoke during pregnancy and did not have a history of hypertension is 6 pounds and 13.01 ounces. Comparatively, mothers who smoked had a average birth weight of 6 pounds and 1.78 ounces, mothers who had a history of hypertension had a average birth weight of 5 pounds and 9.48 ounces, and mothers who smoked and had a history of hypertension had a average birth weight of 5 pounds and 10.34 ounces. This displays that babies whose mothers smoked during pregnancy would have lower average birth weights compared to those who did not smoke. Additionally, the presence of hypertension both with and without smoking has a lower birth weight than those who did not smoke. This analysis reveals that both smoking and hypertension negatively impact birth weight.

## Helpful resources 
Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment inside the code cell, or you can list them below. 

You do not need to cite the official python documentation.
